### PR TITLE
Chart Features and fixes

### PR DESCRIPTION
**Pull request by fur-gan**
- #1568 
- Enhanced crosshair for better orientation in chart view

### CODE REVIEW

File: application/client/src/app/ui/views/toolbar/chart/chart/state.ts [hunk 1]
```diff
@@ -39,8 +45,54 @@ export class State extends AdvancedState {
     }
 
     public onClick(event: MouseEvent) {
+        const position: number = this._calculatePosition(event);
+        position >= 0 && this._session.cursor.select(position, Owner.Chart);
+    }
+
+    public onMouseMove(event: MouseEvent) {
+        const rect: DOMRect = this._element.getBoundingClientRect();
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/toolbar/chart/chart/state.ts:53]: `getBoundingClientRect` is a very-very expensive operation. That's not a good idea to do it with `mousemove` event.

File: application/client/src/app/ui/views/toolbar/chart/chart/state.ts [hunk 2]
```diff
@@ -39,8 +45,54 @@ export class State extends AdvancedState {
...
+        const offsetX = event.offsetX;
+        if (offsetX > 0) {
+            this.crosshairLeft = offsetX;
+            this.showLeftTooltip = offsetX <= rect.width / 2;
+        }
+        if (event.offsetY > 0) {
+            this.tooltipTop = rect.height / 2;
+            const position: number = this._calculatePosition(event);
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/toolbar/chart/chart/state.ts:61]: and inside `this._calculatePosition` one more call of `getBoundingClientRect`.
Be careful with this method it should be used as soon as it's really needed. As usual, we can guess a moment when the size could be changed and at this moment we can measure the size.

File: application/client/src/app/ui/views/toolbar/chart/chart/state.ts [hunk 3]
```diff
@@ -39,8 +45,54 @@ export class State extends AdvancedState {
...
+            })
+            .catch((err: Error) => {
+                this._parent.log().error(err);
+            })
+            .finally(() => {
+                this._loading = false;
+            });
+    }
+
+    private _calculatePosition(event: MouseEvent): number {
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/toolbar/chart/chart/state.ts:93]: In general, this method isn't trustable. Could it be NaN? or infinitive? If it's possible, I would use as output `number | Error` to make it more predictable

**Review by DmitryAstafyev — CHANGES_REQUESTED** [review]

File: application/client/src/app/ui/views/toolbar/chart/chart/state.ts (hunk 3 above)
Comment (fur-gan) [reply application/client/src/app/ui/views/toolbar/chart/chart/state.ts:93]: I'm not sure at which place we could get something else than a number. I mean I make sure that we always calculate the position without failure, or did I miss something?

Comment (fur-gan) [reply application/client/src/app/ui/views/toolbar/chart/chart/state.ts:93]: I see it now, yeah I've changed the code accordingly.

**Pull request merged**